sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -214,25 +448,42 @@ object Hash {
     def addTypedValue(value: Value): this.type
 
     def addCid(cid: Value.ContractId): this.type =
-      addBytes(cid2Bytes(cid))
-
-    private val md = MessageDigestPrototype.Sha256.newDigest
+      addBytes(cid2Bytes(cid), s"${cid.coid} (contractId)")
 
-    override protected def update(a: ByteBuffer): Unit =
+    override protected def update(a: ByteBuffer, context: => String): Unit = {
+      hashTracer.trace(a, context)
       md.update(a)
+    }
 
-    override protected def update(a: Array[Byte]): Unit =
+    override protected def update(a: Array[Byte], context: => String): Unit = {
+      hashTracer.trace(a, context)
       md.update(a)
+    }
 
     override protected def doFinal(buf: Array[Byte]): Unit =
       assert(md.digest(buf, 0, underlyingHashLength) == underlyingHashLength)
 
-    md.update(version.id)
-    md.update(purpose.id)
+    def addVersion: this.type = {
+      addByte(version.id, s"${formatByteToHexString(version.id)} (Value Encoding Version)")
+        .addByte(purpose.id, s"${formatByteToHexString(purpose.id)} (Value Encoding Purpose)")
+    }
   }
 
-  private final class LegacyBuilder(purpose: Purpose, cid2Bytes: Value.ContractId => Bytes)
-      extends ValueHashBuilder(version = Version.Legacy, purpose = purpose, cid2Bytes = cid2Bytes) {
+  /** @param hashVersionAndPurpose Set to false to prevent the value builder from injecting its version and purpose in the hash.
+    *                              This should only be used to get the un-prefixed hash of a single value.
+    */
+  private[crypto] sealed class LegacyBuilder(
```

Comment:
can you add a comment what "Legacy" refers to here

Reply:
or add a todo for me.